// --------------------------------------------------------
// WARNING: XVIZ Protobuf files are unstable, do not use
// --------------------------------------------------------

syntax = "proto3";

package xviz.v2;

option csharp_namespace = "xviz.V2";
option go_package = "v2pb";
option java_multiple_files = true;
option java_outer_classname = "SessionProto";
option java_package = "com.xviz.v2";
option objc_class_prefix = "XVIZ";
option php_namespace = "Xviz\\V2";

import "google/protobuf/struct.proto";
import "xviz/v2/core.proto";
import "xviz/v2/options.proto";
import "xviz/v2/style.proto";

// NOTE: this enum purposely not scoped inside Start to avoid naming conflict
// with "log" field and "log" type
enum SessionType {
  SESSION_TYPE_INVALID = 0;
  live = 1;
  log = 2;
  unbuffered_log = 3;
}

message Start {
  option (xviz_json_schema) = "session/start";
  enum MessageFormat {
    START_MESSAGE_FORMAT_INVALID = 0;
    json = 1;
    binary = 2;
  }
  // XVIZ protocol version, currently 2.0.0
  string version = 1;
  string profile = 2;
  SessionType session_type = 3;
  MessageFormat message_format = 4;
  string log = 5;
}

message TransformLog {
  option (xviz_json_schema) = "session/transform_log";
  string id = 1;
  double start_timestamp = 2;
  double end_timestamp = 3;
  repeated string desired_streams = 4;
}

message StateUpdate {
  option (xviz_json_schema) = "session/state_update";
  enum UpdateType {
    STATE_UPDATE_UPDATE_TYPE_INVALID = 0;
    snapshot = 1; // DO NOT USE - same as incremental
    incremental = 2;
    complete_state = 3;
  }
  UpdateType update_type = 1;
  repeated StreamSet updates = 2;
}

message TransformLogDone {
  option (xviz_json_schema) = "session/transform_log_done";
  string id = 1;
}

message TransformPointInTime {
  option (xviz_json_schema) = "session/transform_point_in_time";
  string id = 1;
  double query_timestamp = 2;
  repeated string desired_streams = 3;
}

message Reconfigure {
  option (xviz_json_schema) = "session/reconfigure";
  enum UpdateType {
    RECONFIGURE_UPDATE_TYPE_INVALID = 0;
    delta = 1;
    full = 2;
  }
  UpdateType update_type = 1;
  google.protobuf.Struct config_update = 2;
}

message Metadata {
  option (xviz_json_schema) = "session/metadata";
  string version = 1;
  map<string, StreamMetadata> streams = 2;
  map<string, CameraInfo> cameras = 3;
  map<string, string> stream_aliases = 4;
  map<string, UIPanelInfo> ui_config = 5;
}

message Error {
  option (xviz_json_schema) = "session/error";
  string message = 1;
}

message StreamMetadata {
  option (xviz_json_schema) = "session/stream_metadata";
  // Separate broad types of XVIZ data
  enum Category {
    STREAM_METADATA_CATEGORY_INVALID = 0;
    primitive = 1;
    time_series = 2;
    variable = 3;
    annotation = 4;
    future_instance = 5;
    pose = 6;
    ui_primitive = 7;
  }
  // The specific data type for time_series and varialbe categories
  enum ScalarType {
    STREAM_METADATA_SCALAR_TYPE_INVALID = 0;
    float = 1;
    int32 = 2;
    string = 3;
    bool = 4;
  }
  // The specific data type for the primitive category
  enum PrimitiveType {
    STREAM_METADATA_PRIMITIVE_TYPE_INVALID = 0;
    circle = 1;
    image = 2;
    point = 3;
    polygon = 4;
    polyline = 5;
    stadium = 6;
    text = 7;
  }
  // The specific type for the ui primitive category
  enum UIPrimitiveType {
    STREAM_METADATA_UI_PRIMITIVE_TYPE_INVALID = 0;
    treetable = 1;
  }
  // The specific type for annotation category
  enum AnnotationType {
    STREAM_METADATA_ANNOTATION_TYPE_INVALID = 0;
    visual = 1;
  }
  enum CoordinateType {
    STREAM_METADATA_COORDINATE_TYPE_INVALID = 0;
    GEOGRAPHIC = 1;
    IDENTITY = 2;
    DYNAMIC = 3;
    VEHICLE_RELATIVE = 4;
  }
  string source = 1;
  string units = 2;
  Category category = 3;
  ScalarType scalar_type = 4;
  PrimitiveType primitive_type = 5;
  UIPrimitiveType ui_primitive_type = 6;
  AnnotationType annotation_type = 7;
  StyleStreamValue stream_style = 8;
  repeated StyleClass style_classes = 9;
  CoordinateType coordinate = 10;
  // 4x4 matrix flattened into 16 elements
  repeated double transform = 11;
  string transform_callback = 12;
}

message CameraInfo {
  option (xviz_json_schema) = "session/camera_info";
  string human_name = 1;
  string source = 2;
  // (x,y,z) vehicle position as 3 elements
  repeated double vehicle_position = 3;
  // 3x3 matrix, flattened into 9 elements
  repeated double vehicle_orientation = 4;
  double pixel_width = 5;
  double pixel_height = 6;
  // 3x3 matrix, flattened into 9 elements
  repeated double rectification_projection = 7;
  // 4x4 matrix, flattened into 9 elements
  repeated double gl_projection = 8;
}

message UIPanelInfo {
  option (xviz_json_schema) = "session/ui_panel_info";
  string name = 1;
  repeated string needed_streams = 2;
  string config = 3;
}
